File: modules/crash-reporting/android/src/main/java/expo/modules/crashreporting/crashEvent.kt
package expo.modules.crashreporting

import io.sentry.SentryEvent
import io.sentry.protocol.Device
import io.sentry.protocol.OperatingSystem
import io.sentry.protocol.SentryStackTrace

private val identifier = Regex("^[a-zA-Z_$][a-zA-Z0-9_$.-]{0,99}$")
private val symbol = Regex("^[a-zA-Z0-9_$ .:<>()\\[\\]+*~&,-]{1,200}$")
private val sourceName = Regex("^[a-zA-Z0-9_.-]{1,200}$")
private fun cleanIdentifier(value: String?) = value?.takeIf { identifier.matches(it) }
private fun cleanSymbol(value: String?) = value?.takeIf { symbol.matches(it) }
private fun fileName(value: String?) = value?.substringBefore('?')?.substringBefore('#')
  ?.substringAfterLast('/')?.substringAfterLast('\\')?.takeIf { sourceName.matches(it) }

private fun cleanStack(stack: SentryStackTrace?) {
  stack?.unknown = null
  stack?.frames?.forEach { frame ->
    frame.vars = null
    frame.preContext = null
    frame.postContext = null
    frame.contextLine = null
    frame.absPath = null
    frame.filename = fileName(frame.filename)
    frame.`package` = fileName(frame.`package`)
    frame.module = cleanIdentifier(frame.module)
    frame.function = cleanSymbol(frame.function)
    frame.rawFunction = cleanSymbol(frame.rawFunction)
    frame.symbol = cleanSymbol(frame.symbol)
    frame.unknown = null
  }
}

internal fun sanitizeCrashEvent(source: SentryEvent): SentryEvent? {
  if (source.exceptions?.any { it.type == "JavascriptException" } == true) return null
  val event = SentryEvent()
  event.eventId = source.eventId
  event.timestamp = source.timestamp
  event.level = source.level
  event.platform = source.platform
  event.release = source.release
  event.dist = source.dist
  event.environment = source.environment
  event.sdk = source.sdk
  event.tags = mapOf("event.origin" to "native", "event.platform" to "android")
  source.contexts.operatingSystem?.let { original ->
    event.contexts.operatingSystem = OperatingSystem().apply {
      name = original.name
      version = original.version
      build = original.build
      kernelVersion = original.kernelVersion
    }
  }
  source.contexts.device?.let { original ->
    event.contexts.device = Device().apply {
      family = original.family
      model = original.model
      modelId = original.modelId
      archs = original.archs
      memorySize = original.memorySize
      isSimulator = original.isSimulator
    }
  }
  event.exceptions = source.exceptions
  event.exceptions?.forEach { exception ->
    exception.type = cleanIdentifier(exception.type) ?: "NativeError"
    exception.value = "Error details omitted for privacy"
    exception.module = cleanIdentifier(exception.module)
    exception.unknown = null
    exception.mechanism?.let {
      it.description = null
      it.data = null
      it.helpLink = null
      it.meta = null
      it.unknown = null
    }
    cleanStack(exception.stacktrace)
  }
  event.threads = source.threads
  event.threads?.forEach {
    it.name = null
    it.unknown = null
    cleanStack(it.stacktrace)
  }
  event.debugMeta = source.debugMeta
  event.debugMeta?.unknown = null
  event.debugMeta?.images?.forEach {
    it.codeFile = fileName(it.codeFile)
    it.debugFile = fileName(it.debugFile)
    it.unknown = null
  }
  return event
}

File: modules/crash-reporting/ios/crashEvent.swift
import Foundation
import Sentry

private func identifier(_ value: String?) -> String? {
  guard let value,
    value.range(of: "^[a-zA-Z_$][a-zA-Z0-9_$.-]{0,99}$", options: .regularExpression) != nil
  else { return nil }
  return value
}

private func symbol(_ value: String?) -> String? {
  guard let value,
    value.range(of: "^[a-zA-Z0-9_$ .:<>()\\[\\]+*~&,-]{1,200}$", options: .regularExpression) != nil
  else { return nil }
  return value
}

private func fileName(_ value: String?) -> String? {
  guard let value else { return nil }
  let path = value.components(separatedBy: CharacterSet(charactersIn: "?#"))[0]
  let name = (path as NSString).lastPathComponent
  return name.range(of: "^[a-zA-Z0-9_.-]{1,200}$", options: .regularExpression) != nil ? name : nil
}

private func cleanStack(_ stack: SentryStacktrace?) {
  for frame in stack?.frames ?? [] {
    frame.vars = nil
    frame.contextLine = nil
    frame.preContext = nil
    frame.postContext = nil
    frame.fileName = fileName(frame.fileName)
    frame.package = fileName(frame.package)
    frame.module = identifier(frame.module)
    frame.function = symbol(frame.function)
  }
}

func sanitizeCrashEvent(_ source: Event) -> Event? {
  // React Native already reports these through its JavaScript error handler.
  if source.exceptions?.contains(where: {
    $0.type.contains("Unhandled JS Exception")
      || $0.value.contains("ExceptionsManager.reportException")
  }) == true {
    return nil
  }

  let event = Event(level: source.level)
  event.eventId = source.eventId
  event.timestamp = source.timestamp
  event.platform = source.platform
  event.releaseName = source.releaseName
  event.dist = source.dist
  event.environment = source.environment
  event.sdk = source.sdk
  event.tags = ["event.origin": "native", "event.platform": "ios"]

  var context: [String: [String: Any]] = [:]
  for (key, fields) in [
    "os": ["name", "version", "build", "kernel_version"],
    "device": ["family", "model", "model_id", "arch", "simulator", "memory_size"],
  ] {
    if let original = source.context?[key] {
      context[key] = original.filter { fields.contains($0.key) }
    }
  }
  event.context = context
  event.exceptions = source.exceptions
  for exception in event.exceptions ?? [] {
    exception.type = identifier(exception.type) ?? "NativeError"
    exception.value = "Error details omitted for privacy"
    exception.module = identifier(exception.module)
    exception.mechanism?.desc = nil
    exception.mechanism?.data = nil
    exception.mechanism?.helpLink = nil
    exception.mechanism?.meta?.error?.domain =
      identifier(exception.mechanism?.meta?.error?.domain) ?? "NativeError"
    cleanStack(exception.stacktrace)
  }
  event.threads = source.threads
  for thread in event.threads ?? [] {
    thread.name = nil
    cleanStack(thread.stacktrace)
  }
  event.stacktrace = source.stacktrace
  cleanStack(event.stacktrace)
  event.debugMeta = source.debugMeta
  for image in event.debugMeta ?? [] {
    image.name = fileName(image.name)
    image.codeFile = fileName(image.codeFile)
  }
  return event
}
